### ISSUE TITLE
[Tech debt] Refine testing of GLM-4.7-Flash and Ling-3.0-Tiny

### ISSUE THREAD

**Issue by mratsim** [mratsim/tattletale#103]
Follow-up to #101, LLM generated. We didn't ship all the tests for:
- https://huggingface.co/zai-org/GLM-4.7-Flash
- https://huggingface.co/inclusionAI/Ling-3.0-tiny

as they fail, some for 1 bf16 ulp. Rather than doing a per-family / per-model / per-case exception I'd rather understand the failure mode properly to refine the statistical validation tests.

The model do run and give coherent replies.

---

## Ship state

Both checkpoints produce coherent output and are usable.
Both still fail fine-grained suites, and each failure is a band or instrument question with an open ruling.

### Ling-3.0-tiny, usable

- Greedy decode on Metal answers correctly in fluent English over 96 tokens at **26.87 tok/s**.
- The 15 GB checkpoint loaded in **2.9 s**.
- No token soup, no repetition loop, tail grammatical, cut only by the token budget.
- The device was printed as `metal` by `deviceName(model.getDeviceKind())` rather than assumed.
- Greedy decoding samples the argmax, so it hides sub-dominant logit error.
- Coherent output rules out a broken port and does not settle a band breach.

### GLM-4.7-Flash, usable

- The 03 suite `t_bf16_glm47flash_03_full_forward_to_logits` passes rc=0 on this box.
- That covers the loader, the MLA pool geometry, the head slice and the rope wiring.

## Failing suites

| Suite | Site | Instrument (frame record) | Kind | Depth | Device | Allowance | Measured drift |
|---|---|---|---|---|---|---|---|
| `t_bf16_ling3_01_layer_internals_mla.nim` | `:68` | `sin_rows` (`tables-Ling-3.0-tiny-00`) | `kElementwise` | 1 | Metal | 2 ulps | **3 ulps** |
| `t_bf16_ling3_01_layer_internals_mla.nim` | `:196` | `output`, prefill seq4 mixer (`attn-Ling-3.0-tiny-00`) | `kReduction` | 1 | CPU | 4 ulps = 0.0625 | **13.5 ulps** = 0.2109375 |
| `t_bf16_glm47flash_04_greedy_text_generation.nim` | per decode step | far-tail probability and truncated KL | `kReduction` | 190 | both | 0.2624 logit units | 1-3 steps per chain breach at top-32 drift of sub-ulp to 2 ulps |

The GLM 04 suite is unregistered in `config.nims` and uncommitted along with its fixtures.

### Band arithmetic

Bands are derived from the recorded magnitudes and never serialized.

- `deriveBands` in `workspace/transformers/tests/harness/harness.nim` computes `ulpBand = ceil(allowanceOf(kind) * sqrt(depth) * amp)`.
- `allowanceOf(kElementwise)` is 2 and `allowanceOf(kReduction)` is 4.
- `amp` is one ulp at the recorded magnitude of the frame record under test.
- `sin_rows.max_magnitude = 0.99997407` sits in binade [0.5, 1) where one f32 ulp is 2^-24 = 5.960464477539e-08.
- The `:68` band is 2 ulps = 1.1920928955078125e-07 and the measured drift is 3 ulps = 1.7881393432617188e-07.
- The sibling `cos_rows.max_magnitude` is exactly 1.0, one binade up, so the two tables in one record reference different ulp sizes.
- The `:196` frame record is `attn-Ling-3.0-tiny-00` with `output.max_magnitude = 3.92188`.
- That magnitude sits in binade [2, 4) where the bf16 ulp is 2^-6 = 0.015625.
- The `:196` band is 4 x 0.015625 = **0.0625** against a measured drift of **0.2109375**, so 13.5 ulps and 3.4x the band.
- The three attn records of this suite carry bands 0.0625, 0.03125 and 0.125, so a band quote without its frame record is meaningless.

## GLM-4.7-Flash 04 decision faithfulness

The suite fails the far-tail instrument while the decisions it exists to test hold.
Evidence recorded in `.scratchspace/20260916-kimi-04-build/salvage-20260917/implement-report-M2.md`.

- On every tail-failing step the **argmax agrees**.
- Chain "The capital of France is" step 30 gives observed id 576 against recorded id 576.
- That step carries top-1 133.0 against recorded 132.0 and rank-32 logit 124.5 against 124.5.
- All pick divergences in the whole run occur at steps whose recorded top-2 margin is exactly **0.0000**.
- A recorded margin of 0.0000 is an exact tie, governed by the harness tie rule.
- Truncated KL sits at **5e-4 to 2.1e-2** against a KL band of about **371-392**.
- Mean signed `ln(observedTail / recordedTail)` per chain is -0.028 to +0.027 on CPU and -0.010 to +0.015 on Metal.
- The drift therefore has **no consistent sign**.
- A real head, norm or scale defect carries a consistent sign.
- The recorded greedy token stream reproduced.

### Why the far-tail instrument is grid-unsatisfiable here

- `tailLimit` caps relative tail movement at 0.3, which is `ln(1.3)` = **0.2624 logit units**.
- On the bf16 grid `ulp(|top1|) >= 0.5` once `|top1| >= 67.2`.
- GLM-4.7-Flash is the first 04 checkpoint whose recorded top-1 logits cross that, with per-chain maxima 144.0, 150.0 and 132.0, so ulp 1.0.
- The smallest possible nonzero disagreement between two honest rows already moves tail probability past the cap, at any depth >= 1.
- No depth or reduction-kind choice changes that band, so the instrument cannot be satisfied for this checkpoint.
- Every other 04 family sits under the threshold, which is why they hold.
- Max recorded `|top1|` is 25.1 on Kimi-Linear, 22.5 on Qwen3-0.6B, 25.5 on Qwen3.5-0.8B and 41.0 on Qwen3.6-35B-A3B.
- One grid step moves tail probability 13.3-28.4% on those checkpoints, all under the cap.
- The chain "Hello how are you" sits exactly at the `MaxTieFlips = 4` ceiling on a 1-thread CPU replay.
- That chain has no tie-flip headroom once the tail question is settled.

## Unresolved

- The 13.5-ulp Ling mixer breach has **never been observed on the reference device**.
- The Metal run dies at `:68`, three assertions before the mixer, so the mixer carries no Metal measurement.
- No claim that the mixer is broken, or is not broken, on Metal is evidenced until that measurement exists.
- Getting that measurement requires getting past the sin-table assertion, which is the kind ruling below.
- Whether the 3-ulp `sin_rows` overshoot is a libm or evaluation-order difference or a real defect is undetermined.
- A one-ulp step past a single-stage elementwise allowance has the shape of an evaluation-order difference.
- That shape is evidence, not a verdict.
- The far-tail band derivation for high-magnitude logits is open.
- One proposed shape is to floor the relative tail allowance at one grid step of the recorded magnitude instead of capping it at 0.3.
- That proposal edits `workspace/transformers/tests/harness/`, needs an operator ruling, and loosens a band.

## Reference device

- The GPU is the reference device, Metal on macOS and CUDA on Linux.
- CPU is a degraded fallback, and a CPU result is not evidence about the port.
- `PYTORCH_ENABLE_MPS_FALLBACK` **permits** silent CPU fallback where MPS lacks a kernel.
- It does not force ops to CPU, because PyTorch still prefers the MPS kernel when one exists.
- A pass with the flag set proves nothing about which device produced the numbers.
- With the variable absent an op that MPS lacks hard-errors, so a pass without it proves every dispatched op has an MPS kernel.
- Five of the six committed 04 suites pass without the flag.
- They are `qwen3_04`, `qwen35dense_04`, `qwen36moe_04`, `kimi_04` on 92 GB and `moonlight_04`.
- `exl3_qwen3_04` is unverified, killed past a 300 s window having reached load and one full assert-pass with zero missing-kernel errors.
- The EXL3 5bpw dequant-to-matmul chain crawls on MPS next to its bf16 sibling, which is its own card and orthogonal to this one.
- Seven committed suite files still set the flag in their own source, which makes fallback silent rather than necessary.
- Each removal must also fix the comment above it claiming Metal falls back to CPU kernels, which becomes false after removal.

## Non-goals

- Do not widen a band, reclassify a kind or drop an instrument to make a suite pass.
- Band widening requires an operator ruling, per J139, J151 and J187, "never loosen to accommodate drift, escalate".
- A band that moves to fit an observation stops being a measurement.
- Do not key the far-tail allowance to the recorded answer.
- Do not let a device, box, thread count or recorded-environment entry select a branch.
- The aggregate skip of the Ling MLA suite is not a resolution, it is a way to keep the aggregate run honest while the rulings are open.

## Open rulings

1. Is a RoPE theta table an elementwise or a reduction perturbation?
   The table is built from `cos` and `sin` of per-position products.
   `kElementwise` gives a 2-ulp allowance, `kReduction` gives 4, and the observed drift is 3.
   The class must be chosen on what the op is, and the looser class is not selectable because it fits the observation.
2. How is the far-tail allowance derived for high-magnitude logits without keying it to the recorded answer?
   The derivation must stay satisfiable on the bf16 grid at `|top1| >= 67.2`.
   It must stay sensitive to a real tail defect at the magnitudes where the current cap is reachable.

## Handling in the tree

- `config.nims` carries `AggregateSkippedSuites`, and `runFamily` echoes the suite name and the reason on every skip.
- The aggregate paths `test_transformers` and every `test_tf_family` skip `t_bf16_ling3_01_layer_internals_mla.nim`.
- The suite still runs in full through `nim test_tf_bf16_ling3_01_layer_internals_mla`.
- The skip is an aggregate convenience and not a way to hide work.
- The GLM-4.7-Flash 04 fixtures and suite are untracked and the suite is unregistered, so it cannot block a build or a merge.
- Each failing Ling site carries a one-line `# TODO` naming the instrument, kind, allowance and measured drift.

## Reproduce

```bash
# One suite, device selected by TTT_TEST_ON
nim test_tf_bf16_ling3_01_layer_internals_mla
TTT_TEST_ON=cpu nim test_tf_bf16_ling3_01_layer_internals_mla
TTT_TEST_ON=metal nim test_tf_bf16_ling3_01_layer_internals_mla

# Reference-device run, CPU fallback disallowed
env -u PYTORCH_ENABLE_MPS_FALLBACK nim test_tf_bf16_qwen36moe_04_greedy_text_generation

# GLM 04 suite is unregistered, so it runs as a direct compile
nim cpp -r -d:release --stackTrace:on --lineTrace:on --lineDir:on --debugger:native \
  --hints:off --warnings:off --passC:"-std=c++20" \
  --outdir:build/tests --nimcache:nimcache/tests \
  workspace/transformers/tests/q_bf16/t_bf16_glm47flash_04_greedy_text_generation.nim

# Aggregate skip, loud, no compile
TTT_TEST_ON=junk nim test_tf_family name=ling3
```

Use a warm nimcache, `--nimcache:nimcache/wip` for scratch builds.
A cold path re-semchecks the libtorch graph and costs minutes.
Check `pgrep -fl "nim cpp"` before compiling, the nimcache is shared with other agents.
